File: Chapter15/TestAlarmManager/Java/TestAlarmManager/app/src/main/java/com/androidbook/alarms/TestAlarmsDriverActivity.java
package com.androidbook.alarms;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.TextView;

public class TestAlarmsDriverActivity extends AppCompatActivity implements IReportBack {

    private static final String TAG = TestAlarmsDriverActivity.class.getSimpleName();

//    private SendAlarmOnceTester alarmTester = null;
//    private SendRepeatingAlarmTester alarmTester = null;
//    private CancelRepeatingAlarmTester alarmTester = null;
//    private ScheduleIntentMultipleTimesTester alarmTester = null;
    private AlarmIntentPrimacyTester alarmTester = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        alarmTester = new AlarmIntentPrimacyTester(this, this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        MenuInflater inflater = getMenuInflater();  // from activity
        inflater.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        appendMenuItemText(item);
        switch (item.getItemId()) {
            case R.id.menu_clear:
                emptyText();
                break;

            case R.id.menu_alarm_once:
                alarmTester.sendAlarmOnce();
                break;

            case R.id.menu_alarm_repeated:
                alarmTester.sendRepeatingAlarm();
                break;

            case R.id.menu_alarm_cancel:
                alarmTester.cancelRepeatingAlarm();
                break;

            case R.id.menu_alarm_multiple:
                alarmTester.scheduleSameIntentMultipleTimes();
                break;

            case R.id.menu_alarm_distinct_intents:
                alarmTester.scheduleDistinctIntents();
                break;

            case R.id.menu_alarm_intent_primacy:
                alarmTester.alarmIntentPrimacy();
                break;
        }
        return true;
    }

    private TextView getTextView() {
        return findViewById(R.id.text1);
    }

    private void appendMenuItemText(MenuItem item) {
        String title = item.getTitle().toString();
        TextView tv = getTextView();
        tv.setText(String.format("%s\n%s", tv.getText(), title));
    }

    private void emptyText() {
        TextView tv = getTextView();
        tv.setText("");
    }

    private void appendText(String s) {
        TextView tv = getTextView();
        tv.setText(String.format("%s\n%s", tv.getText(), s));
        Log.d(TAG, s);
    }

    @Override
    public void reportBack(String tag, String message) {
        appendText(tag + ":" + message);
        Log.d(TAG, message);
    }
}

File: Chapter15/TestAlarmManager/Kotlin/TestAlarmManager/app/src/main/java/com/androidbook/alarms/TestAlarmsDriverActivity.kt
package com.androidbook.alarms

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView

class TestAlarmsDriverActivity : AppCompatActivity(), IReportBack {

//    private lateinit var alarmTester: SendAlarmOnceTester
//    private lateinit var alarmTester: SendRepeatingAlarmTester
//    private lateinit var alarmTester: CancelRepeatingAlarmTester
//    private lateinit var alarmTester: ScheduleIntentMultipleTimesTester
    private lateinit var alarmTester: AlarmIntentPrimacyTester

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        alarmTester = AlarmIntentPrimacyTester(this, this)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        appendMenuItemText(item)
        when (item.itemId) {
            R.id.menu_clear -> emptyText()
            R.id.menu_alarm_once -> alarmTester.sendAlarmOnce()
            R.id.menu_alarm_repeated -> alarmTester.sendRepeatingAlarm()
            R.id.menu_alarm_cancel -> alarmTester.cancelRepeatingAlarm()
            R.id.menu_alarm_multiple -> alarmTester.scheduleSameIntentMultipleTimes()
            R.id.menu_alarm_distinct_intents -> alarmTester.scheduleDistinctIntents()
            R.id.menu_alarm_intent_primacy -> alarmTester.alarmIntentPrimacy()
        }
        return true
    }

    @SuppressLint("WrongViewCast")
    private fun getTextView(): TextView {
        return findViewById(R.id.text1)
    }

    private fun appendMenuItemText(item: MenuItem) {
        val tv = getTextView()
        tv.text = String.format("%s\n%s", tv.text, item.title)
    }

    private fun emptyText() {
        getTextView().text = ""
    }

    private fun appendText(s: String) {
        val tv = getTextView()
        tv.text = String.format("%s\n%s", tv.text, s)
    }

    override fun reportBack(tag: String, message: String) {
        appendText("$tag:$message")
        Log.d(tag, message)
    }

}